Modules/_dbmmodule.c

```diff
@@ -78,14 +86,22 @@ dbm_dealloc(dbmobject *dp)
 {
     if ( dp->di_dbm )
         dbm_close(dp->di_dbm);
-    PyObject_Del(dp);
+    PyTypeObject *tp = Py_TYPE(dp);
+    freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
+    free_func(dp);
+    Py_DECREF(tp);
 }
 
 static Py_ssize_t
 dbm_length(dbmobject *dp)
 {
+    PyTypeObject *tp = Py_TYPE(dp);
+    _dbm_state *state =  PyType_GetModuleState(tp);
```

Comment:
I would prefer `assert(state != NULL);` since this call **cannot** fail with the current code (see my comment about Py_TPFLAGS_BASETYPE).

For example (master branch, without this change):

```
Python 3.10.0a0 (heads/master:25f38d7044, Jun 15 2020, 14:16:18) 
>>> import _dbm
>>> d=_dbm.open('test.db', "c")
>>> dbm_type = type(d)

# legit
>>> dbm_type.setdefault(d, "key", "value")
'value'

# invalid: error
>>> dbm_type.setdefault("string", "key", "value")
TypeError: descriptor 'setdefault' for '_dbm.dbm' objects doesn't apply to a 'str' object
```

It seems like the descriptor checks the "self" type.